src/rubrix/labeling/text_classification/weak_labels.py

```diff
@@ -1088,6 +1125,57 @@ def show_records(
 
         return pd.DataFrame(map(lambda x: x.dict(), filtered_records))
 
+    def extend_matrix(
+        self,
+        thresholds: Union[List[float], np.ndarray],
+        embeddings: Optional[np.ndarray] = None,
+        gpu: bool = False,
+    ):
+        """Extends the weak label matrix through embeddings according to the similarity thresholds for each rule.
+           Implementation based on `Epoxy <https://github.com/HazyResearch/epoxy>`__.
+
+        Args:
+            thresholds: An array of thresholds between 0.0 and 1.0, one for each column of the weak labels matrix.
+                Each one stands for the minimum cosine similarity between two sentences for a rule to be extended.
+            embeddings: Embeddings for each row of the weak label matrix.
+                If not provided, we will use the ones from the last ``WeakMultiLabels.extend_matrix()`` call.
+            gpu: If True, perform FAISS similarity queries on GPU.
+
+        Examples:
+            >>> # Choose any model to generate the embeddings.
+            >>> from sentence_transformers import SentenceTransformer
+            >>> model = SentenceTransformer('all-mpnet-base-v2', device='cuda')
+            >>>
+            >>> # Generate the embeddings and set the thresholds.
+            >>> weak_labels = WeakMultiLabels(dataset="my_dataset")
+            >>> embeddings = np.array([ model.encode(rec.text) for rec in weak_labels.records() ])
+            >>> thresholds = [0.6] * len(weak_labels.rules)
+            >>>
+            >>> # Extend the weak labels matrix.
+            >>> weak_labels.extend_matrix(thresholds, embeddings)
+            >>>
+            >>> # Calling the method below will now retrieve the extended matrix.
+            >>> weak_labels.matrix()
+            >>>
+            >>> # Subsequent calls without the embeddings parameter will reutilize the faiss index built on the first call.
+            >>> thresholds = [0.75] * len(weak_labels.rules)
+            >>> weak_labels.extend_matrix(thresholds)
+            >>> weak_labels.matrix()
+        """
+        support = [
+            np.argwhere(self._matrix[:, i].sum(-1) >= 0).flatten()
+            for i in range(self._matrix.shape[1])
+        ]
+
+        abstains = [
+            np.argwhere(self._matrix[:, i].sum(-1) < 0).flatten()
+            for i in range(self._matrix.shape[1])
+        ]
+
+        self._extended_matrix = self._extend_matrix(
+            thresholds, embeddings, abstains, support, gpu
+        )
```

Comment:
Since both, single and multi-label version of this method are almost the same, could we do some refactors to create a common base method with separate internal steps? 

Something like that:

```python
def extend_matrix(
        self,
        thresholds: Union[List[float], np.ndarray],
        embeddings: Optional[np.ndarray] = None,
        gpu: bool = False,
    ):
    
    support, abstain = self._extend_matrics_preprocess()
    extended_matrix = self._extend_matrix(
            thresholds, embeddings, abstains, support, GPU
        )
    self._extended_matrix = self._extend_matrix_postprocess(extended_matrix)
```

then, every weak label child class can implement its methods.

Reply:
Done